ledger/simulation/simulation_eval_test.go

```diff
@@ -849,6 +866,152 @@ int 0
+		expensiveTxn := txnInfo.NewTxn(txntest.Txn{
+			Type:          protocol.ApplicationCallTx,
+			Sender:        sender.Addr,
+			ApplicationID: futureAppID,
+			Accounts:      []basics.Address{receiver.Addr},
+		})
+
+		txntest.Group(&createTxn, &expensiveTxn)
+
+		signedCreateTxn := createTxn.Txn().Sign(sender.Sk)
+		signedExpensiveTxn := expensiveTxn.Txn().Sign(sender.Sk)
+
+		return simulationTestCase{
+			input:         []transactions.SignedTxn{signedCreateTxn, signedExpensiveTxn},
+			expectedError: "dynamic cost budget exceeded",
+			expected: simulation.Result{
+				Version:   1,
+				LastRound: txnInfo.LatestRound(),
+				TxnGroups: []simulation.TxnGroupResult{
+					{
+						Txns: []simulation.TxnResult{
+							{
+								Txn: transactions.SignedTxnWithAD{
+									ApplyData: transactions.ApplyData{
+										ApplicationID: futureAppID,
+									},
+								},
+								BudgetUsed: 4,
+							},
+							{
+								BudgetUsed: 1396,
+							},
+						},
+						FailedAt:       simulation.TxnPath{1},
+						BudgetAdded:    1400,
+						BudgetConsumed: 1400,
+					},
+				},
+				WouldSucceed: false,
+			},
+		}
+	})
+}
+
+func TestLogicSigOverBudget(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	op, err := logic.AssembleString(`#pragma version 6
+` + strings.Repeat(`byte "a"
+keccak256
+pop
+`, 200) + `int 1`)
+	require.NoError(t, err)
+	program := logic.Program(op.Program)
+	lsigAddr := basics.Address(crypto.HashObj(&program))
+
+	simulationTest(t, func(accounts []simulationtesting.Account, txnInfo simulationtesting.TxnInfo) simulationTestCase {
+		sender := accounts[0]
+
+		payTxn := txnInfo.NewTxn(txntest.Txn{
+			Type:     protocol.PaymentTx,
+			Sender:   sender.Addr,
+			Receiver: lsigAddr,
+			Amount:   1_000_000,
+		})
+		appCallTxn := txnInfo.NewTxn(txntest.Txn{
+			Type:   protocol.ApplicationCallTx,
+			Sender: lsigAddr,
+			ApprovalProgram: `#pragma version 8
+byte "hello"
+log
+int 1`,
+			ClearStateProgram: `#pragma version 8
+int 1`,
+		})
+
+		txntest.Group(&payTxn, &appCallTxn)
+
+		signedPayTxn := payTxn.Txn().Sign(sender.Sk)
+		signedAppCallTxn := appCallTxn.SignedTxn()
+		signedAppCallTxn.Lsig = transactions.LogicSig{
+			Logic: program,
+		}
+
+		var expectedAppCallAD transactions.ApplyData
+		expectedFailedAt := simulation.TxnPath{1}
+
+		return simulationTestCase{
+			input:         []transactions.SignedTxn{signedPayTxn, signedAppCallTxn},
+			expectedError: "dynamic cost budget exceeded",
+			expected: simulation.Result{
+				Version:   1,
+				LastRound: txnInfo.LatestRound(),
+				TxnGroups: []simulation.TxnGroupResult{
+					{
+						Txns: []simulation.TxnResult{
+							{},
+							{
+								Txn: transactions.SignedTxnWithAD{
+									ApplyData: expectedAppCallAD,
+								},
+								BudgetUsed: 0,
+							},
+						},
+						FailedAt:       expectedFailedAt,
+						BudgetAdded:    0,
+						BudgetConsumed: 0,
+					},
+				},
+				WouldSucceed: false,
+			},
+		}
+	})
+}
+
```

Comment:
Based on my first reading of the code, I thought there might be a bug if an app used _exactly_ its budget before invoking an inner txn. The good news is that I was wrong and that scenario works correctly, but I ended up making this test case that's still probably worth including:

```go
func TestAppAtBudget(t *testing.T) {
	partitiontest.PartitionTest(t)
	t.Parallel()

	// Transaction has a cost of 700 and invokes an inner transaction
	exactly700AndCallInner := fmt.Sprintf(`#pragma version 6
pushint 1
cover 0 // This is a noop, just to fix an odd number of ops
%s
itxn_begin
int appl
itxn_field TypeEnum
byte 0x068101
dup
itxn_field ClearStateProgram
itxn_field ApprovalProgram
itxn_submit
`, strings.Repeat(`pushint 1
pop
`, 345))

	simulationTest(t, func(accounts []simulationtesting.Account, txnInfo simulationtesting.TxnInfo) simulationTestCase {
		sender := accounts[0]

		futureAppID := basics.AppIndex(2)
		// fund outer app
		fund := txnInfo.NewTxn(txntest.Txn{
			Type:     protocol.PaymentTx,
			Sender:   sender.Addr,
			Receiver: futureAppID.Address(),
			Amount:   401_000,
		})
		// create app
		appCall := txnInfo.NewTxn(txntest.Txn{
			Type:            protocol.ApplicationCallTx,
			Sender:          sender.Addr,
			ApprovalProgram: exactly700AndCallInner,
			ClearStateProgram: `#pragma version 6
int 1`,
		})

		txntest.Group(&fund, &appCall)

		signedFundTxn := fund.Txn().Sign(sender.Sk)
		signedAppCall := appCall.Txn().Sign(sender.Sk)

		return simulationTestCase{
			input: []transactions.SignedTxn{signedFundTxn, signedAppCall},
			expected: simulation.Result{
				Version:   1,
				LastRound: txnInfo.LatestRound(),
				TxnGroups: []simulation.TxnGroupResult{
					{
						Txns: []simulation.TxnResult{
							{},
							{
								Txn: transactions.SignedTxnWithAD{
									ApplyData: transactions.ApplyData{
										ApplicationID: futureAppID,
										EvalDelta: transactions.EvalDelta{
											InnerTxns: []transactions.SignedTxnWithAD{
												{
													ApplyData: transactions.ApplyData{
														ApplicationID: futureAppID + 1,
													},
												},
											},
										},
									},
								},
								BudgetUsed: 701,
							},
						},
						BudgetAdded:    1400,
						BudgetConsumed: 701,
					},
				},
				WouldSucceed: true,
			},
		}
	})
}
```